RFC: Add 2021-11-24-graceful-restart-tikv-pod.md (#4277)
pingcap/tidb-operator: docs/design-proposals/2021-11-24-graceful-reschedule-tikv-pod.md

# gracefully reschedule tikv pod

TiDB-Operator issue: [#4215](https://github.com/pingcap/tidb-operator/issues/4215)

## Summary

Support gracefully reschedule a tikv pod. The implemtation is necessarrily the combination of three operations:

1. Add a PD evict-leader-scheduler to transfer the leader.
2. When leader count drop down to zero, delete the pod to let it re-create and reschedule.
3. Remove the evict-leader-scheduler.

## Motivation

### Goals

- Provide a way for user to gracefully reschedule a tikv pod.
- The pod must be gracefully stop, to be specify, evict leaders before delete the pod and must work nomal when recreate the pod.

### Non-Goals

- Provider a way to make tikv graceful stop when just recive a `SIGTERM` signal.

## Proposal

### User Stories

#### Story 1

> suppost `tikv-0` is running at node `node0`,  there are many free resource at other nodes, I would like to make tikv-0 run at other node and shutdown `node0` to reduce the cost. I may apply the flowing operations:

Reviewer: ```suggestion
suppose `tikv-0` is running at node `node0`,  there are many free resource at other nodes, I would like to make tikv-0 run at other node and shutdown `node0` to reduce the cost. I may apply the flowing operations:
```

1. run `kubectl cordon node0` to mark `node0`as unschedulable.
2. delete pod `tikv-0` to let it re-create and reschedule to other nodes.
3. draine `node-0` in the nomal way as [safely drain a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) and shotdown `node0`

### Risks and Mitigations

- PD/TiKV fails to evict all leader and we may just keep checking it.

## Design Details

> Support user to add an annotation to CRD `TidbCluster` to trigger an gracefule reschedule.

Reviewer: Why not update pod annotation to reschedule it?

Reviewer: > then we may lose the update(leave an evict-leader scheduler at pd)
While the pod has the annotation, add evict-leader scheduler at pd. If the pod is deleted or updated(annotation is removed), just remove scheduler at pd. It's level triggered and cancelable so scheduler will not be lefted at pd.

Author: You can not always catch the `delete`(like at the time pod deleted operator not running normal), you need to check time from time.

another possible issue is we may also delete evict-leader scheduler not added by us.

Reviewer: > You can not always catch the delete(like at the time pod deleted operator not running normal), you need to check time from time.

No need to catch the event. We just check the current tikv pod. While pod is deleted and operator is not running, delete event will miss but evict-leader scheduler can be removed after operator is restarted.

> another possible issue is we may also delete evict-leader scheduler not added by us.

It seems a common issue for both of two ways?

Author: > No need to catch the event. We just check the current tikv pod. While pod is deleted and operator is not running, delete event will miss but evict-leader scheduler can be removed after operator is restarted.
--

yes, you need to check all clusters as mentioned before. (if not annotation but having an evict-leader scheduler, we delete the scheduler)

> It seems a common issue for both of two ways?

no, we have the annotation at tc, we will not delete the scheduler added by the user.

Reviewer: Fine, I understand your concern. Let me rethink it because I prefer to act it in pod level.

annotation key: `tikv.tidb.pingcap.com/reschedule`

> annotation value is JSON in the flowing struct:

Reviewer 2: This json is too complicated for users to use.

```go
{
  // the flowing field set by user.

>   NodeName *string

Reviewer 3: Why not support specify more than one node?

Reviewer: Why nodeName is needed?

Author: generally, we just want to take one node per time?
you can do it one by one.

Reviewer 3: > generally, we just want to take one node per time?

but if the user set more than one node, operator can still do it one by one, right?

  Slots []int 
  
  // the flowing filed set by controller.
  RunningSlot *RunningSlot
  FinishedSlots []int
  Status string // Running|Finished
  Message string // optial message of the current status for human readible 
}

RunningSlot struct {
  Slot int
  Status string // Evicting|ReCreating
  StartTimestamp metav1.Time // etc. 2021-11-23T13:09:32Z
}
```

One of `NodeName` and `Slots` is required, user is able to specify all pods in a node or specify the pods directly.

At controller side, when ever observe this annotation is set, it will do the flowing step to handle an unfinished action:

```go
if RunningSlot is set {
   switch RunningSlotStatus {
     case "Evicting"
       // 1. make sure evict-scheduler is added
       // 2. update RunningSlot.Status as ReCreating if leader-count is 0
     case "ReCreating"
     if the creationTimestamp of pod is less than RunningSlot.startTimestamp {
        delete the pod.
     }
     if the pod is already running normal {

>         append RunningSlot.Slot in RunningSlot and set RunningSlot as nil

Reviewer 2: `append RunningSlot.Slot in RunningSlot` to `FinishedSlots`?

Reviewer 2: When to delete the evict leader scheduler?

     }
   }
} else {
  // Find a slot to hanle by comaring `FinishedSlots` and `Slots` or slots belong to the specify NodeName.
  // set RunningSlot or mark Status as Finished
}

```

> An example for Story 1 step 2, user might add annotation with the flowing value for key `tikv.tidb.pingcap.com/reschedule`:

Reviewer 3: did the user need to do step 1 (`kubectl cordon node0`) manually? this will also affect the schedule of other pods? can we modify our sts to achieve this goal?

Author: > did the user need to do step 1 (kubectl cordon node0) manually?

yes

I prefer operator does not care anything about the schedule. that means a user can do not do `kubectl cordon node0`
and just let it re-create and schedule in the normal way(can still run at `node0` or other nodes) if the user wants this(like graceful restart in tiup).

Reviewer 2: Yes, the original requirement is just to restart TiKV gracefully.

2. delete pod `tikv-0` to let it re-create and reschedule to other nodes, to mitigate the impact of unavailability of some region, I will do the flowing operations insted of delete the pod directly:
   1. add evict-leader-scheduler by `pd-ctl`.
   2. delete `tikv-0` pod when leader count drop down to 0.
   3. remove the evict-leader-scheduler by `pd-ctl`
3. drain `node-0` in the normal way as [safely drain a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) and shotdown `node0`

### Risks and Mitigations

- PD/TiKV fails to evict all leader and we may just keep checking it.

## Design Details

Support user to add an annotation to tikv pod to trigger an graceful reschedule.

Annotation key `EvictLeaderAnnKey`: `tikv.tidb.pingcap.com/evict-leader`

the value controler the behavier when the leander count drop to zero, valid value is one of:

- `none`: doing nothing
- `delete-pod`: delete pod and remove the evict-leader scheduler from PD.

An EvictLeader status willbe added to the `TiKVStatus`:

```go
+type EvictLeaderStatus struct {
+       PodCreateTime metav1.Time `json:"podCreateTime,omitempty"`

> +       Value         string      `json:"value,omitempty"`

Reviewer 3: do we need to add pod name into this status?

Author: https://github.com/pingcap/tidb-operator/pull/4277/commits/edece78c434d7035b303b470ac8e8f3c38f88862
s/*EvictLeader/map[string]*EvictLeader
in tikv status to support multi pod annotated in the same time.

### Non-Goals

- Provider a way to make tikv graceful stop when just recive a `SIGTERM` signal.

## Proposal

### User Stories

#### Story 1

suppose `tikv-0` is running at node `node0`,  there are many free resource at other nodes, I would like to make tikv-0 run at other node and shutdown `node0` to reduce the cost. I may apply the flowing operations:

1. run `kubectl cordon node0` to mark `node0` as `unschedulable`.
2. delete pod `tikv-0` to let it re-create and reschedule to other nodes, to mitigate the impact of unavailability of some region, I will do the flowing operations insted of delete the pod directly:
   1. add evict-leader-scheduler by `pd-ctl`.
   2. delete `tikv-0` pod when leader count drop down to 0.
   3. remove the evict-leader-scheduler by `pd-ctl`
3. drain `node-0` in the normal way as [safely drain a node](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) and shotdown `node0`

### Risks and Mitigations

- PD/TiKV fails to evict all leader and we may just keep checking it.

## Design Details

Support user to add an annotation to tikv pod to trigger an graceful reschedule.

> Annotation key `EvictLeaderAnnKey`: `tikv.tidb.pingcap.com/evict-leader`

Reviewer 2: What about using two separate keys `tidb.pingcap.com/evict-leader` and `tidb.pingcap.com/restart`?

Author: so the behavior of `tidb.pingcap.com/evict-leader` with any value is the same as a `none` value here.
and `tidb.pingcap.com/restart` is same as a `delete-pod` value here?

Reviewer 2: Yes.

the value controler the behavier when the leander count drop to zero, valid value is one of:

- `none`: doing nothing
- `delete-pod`: delete pod and remove the evict-leader scheduler from PD.

> An EvictLeader status willbe added to the `TiKVStatus`:

Reviewer: ```suggestion
An EvictLeader status will be added to the `TiKVStatus`:
```

```go
+type EvictLeaderStatus struct {

> +       PodCreateTime metav1.Time `json:"podCreateTime,omitempty"`

Reviewer 2: Is this field used anywhere?

+       Value         string      `json:"value,omitempty"`
+}

 // TiKVStatus is TiKV status
 type TiKVStatus struct {
        Synced          bool                            `json:"synced,omitempty"`
@@ -1139,6 +1151,7 @@ type TiKVStatus struct {
        TombstoneStores map[string]TiKVStore            `json:"tombstoneStores,omitempty"`
        FailureStores   map[string]TiKVFailureStore     `json:"failureStores,omitempty"`
        Image           string                          `json:"image,omitempty"`
+       EvictLeader     map[string]*EvictLeaderStatus   `json:"evictLeaderStatus,omitempty"`
 }
```

> At controller side, the reconciler to handle pod will be like:

Reviewer 2: A new controller?

Author: yes

```go
func sync(pod *corev1.Pod, tc *v1alpha1.TidbCluster) (ctrl.Result, error) {
        value, ok := pod.Annotations[v1alpha1.EvictLeaderAnnKey]

        if ok {
                evictStatus := &v1alpha1.EvictLeaderStatus{
                        PodCreateTime: pod.CreationTimestamp,
                        Value:         value,
                }
                nowStatus := tc.Status.TiKV.EvictLeader[pod.Name]

>                 if nowStatus == nil || *nowStatus != evictStatus {

Author: ```suggestion
                if nowStatus == nil || *nowStatus != *evictStatus {
```

2. Delete to the pod to make it re-create when leader-count is 0.
3. Remove the evict-leader-scheduler when the new pod becomes ready.

For the user only want to trigger an evict-leader-scheduler by annotating a pod, we will also support another key for this.

Annotation key:`tidb.pingcap.com/evict-leader`

The controller will do the flowing operations:

1. Add evict-leader-scheduler for the tikv store.

2. Remove the evict-leader-scheduler once is annotation is deleted.

An EvictLeader status will be added to the `TiKVStatus`:

```go
+type EvictLeaderStatus struct {
+       PodCreateTime metav1.Time `json:"podCreateTime,omitempty"`
+       Value         string      `json:"value,omitempty"`
+}

 // TiKVStatus is TiKV status
 type TiKVStatus struct {
        Synced          bool                            `json:"synced,omitempty"`
@@ -1139,6 +1151,7 @@ type TiKVStatus struct {
        TombstoneStores map[string]TiKVStore            `json:"tombstoneStores,omitempty"`
        FailureStores   map[string]TiKVFailureStore     `json:"failureStores,omitempty"`
        Image           string                          `json:"image,omitempty"`

> +       EvictLeader     map[string]*EvictLeaderStatus   `json:"evictLeaderStatus,omitempty"`

Reviewer 2: Is `EvictLeaderStatus` useful here? If not, can we just use `EvictLeaderPods map[string]EmptyStruct `json:"evictLeaderPods,omitempty"``?

 }
```

The `Value` of `EvictLeaderStatus` control the behavior when the leader count drops to zero, the valid value is one of:

- `none`: doing nothing (for `tidb.pingcap.com/evict-leader`)

> - `delete-pod`: delete pod and remove the evict-leader scheduler from PD. (for `tidb.pingcap.com/restart`)

Reviewer 2: Needs to update here.
